### ISSUE TITLE
GCop406 false alarm

### ISSUE THREAD

**Issue by sungam3r** [Geeksltd/GCop#252]
I have GCop406 on all fields which are assigned through `ref` arg.

```
    internal class A : IDisposable
    {
        private long _disposed; // OK

        public void Dispose()
        {
            _disposed = 1;
        }
    }
```

```
    internal class A : IDisposable
    {
        private long _disposed; // GCop406	 Mark _disposed field as read-only.	

        public void Dispose()
        {
            Interlocked.Increment(ref _disposed);
        }
    }
```

Of course, doing this leads to
CS0192	A readonly field cannot be used as a ref or out value (except in a constructor)

**Comment by Foroughi** [comment]
@sungam3r Done. You may wanna wait for the next release. Assuming we have 

`private int foo;`

here is the test results : 

- [X] No usage of `foo` in the current class
- [X] bar(foo);
- [ ] bar(ref foo);
- [ ] bar(out foo);
- [ ] foo = something;


Again thanks for the report

**Comment by marjanjavid** [comment]
It works fine on GCop2.6.0 so I close this issue

**Comment by sungam3r** [comment]
Some of the bugs are gone, but I still have GCop406 false alarms:
```csharp
    internal sealed class AsyncTest
    {
        public AsyncTest(Func<Task> _)
        {
        }
    }

    internal sealed class Test
    {
        private long _counter;  // GCop406	 Mark _counter field as read-only.	

        public Test()
        {
            new AsyncTest(async () =>
            {
                await Task.Delay(10);
                Interlocked.Increment(ref _counter);
            });
        }
    }
```
And also this:

```csharp
 public static class AsyncExtensions
    {
        private const int DEFAULT_TIMEOUT_MSEC = 60000; // GCop406	 Mark DEFAULT_TIMEOUT_MSEC field as read-only.	
       
        public static void GetCacheItemAsync<T>(this string key, int timeout = DEFAULT_TIMEOUT_MSEC)
        {
        }
}
```

**Comment by Foroughi** [comment]
@sungam3r Would you please let me know whats the version your using !?

**Comment by sungam3r** [comment]
```xml
<ItemGroup>
    <PackageReference Include="GCop.All.Common" Version="2.6.0" PrivateAssets="all" />
</ItemGroup>
```
![image](https://user-images.githubusercontent.com/21261007/51114533-a2baa980-1816-11e9-8266-343570e7532f.png)

**Comment by Foroughi** [comment]
This one is fixed. GCop406 wont kick in for the properties which are decorated with `const `modifier. You will have it fixed in the next release 
```
public static class AsyncExtensions
    {
        private const int DEFAULT_TIMEOUT_MSEC = 60000; // GCop406	 Mark DEFAULT_TIMEOUT_MSEC field as read-only.	
       
        public static void GetCacheItemAsync<T>(this string key, int timeout = DEFAULT_TIMEOUT_MSEC)
        {
        }
}
```

**Comment by Foroughi** [comment]
@sungam3r this one is also fixed , you will have it fixed in the next release 

```
 internal sealed class AsyncTest
    {
        public AsyncTest(Func<Task> _)
        {
        }
    }

    internal sealed class Test
    {
        private long _counter;  // GCop406	 Mark _counter field as read-only.	

        public Test()
        {
            new AsyncTest(async () =>
            {
                await Task.Delay(10);
                Interlocked.Increment(ref _counter);
            });
        }
    }
```

**Comment by sungam3r** [comment]
Async lambda may be not only in constructor. My example does not focus on outer context - constructor, it focuses on async lambda itself.

**Comment by Foroughi** [comment]
@sungam3r Normally if the criteria is in a method the rule wont ignore it unless there is a bug with it, Ill have check on it

**Comment by marjanjavid** [comment]
The changes are available on GCop2.6.1. So I close this issue

**Comment by sungam3r** [comment]
Still have this error in 2.6.1

**Comment by sungam3r** [comment]
I am surprised, because I see this error on the most primitive code:
```c#
  internal class Test
    {
        private int counter; // GCop406	Mark counter field as read-only.	
        public void DoSome() => ++counter;
    }
```
